File: mycuration/src/main/java/com/phicdy/mycuration/rss/UnreadCountManager.java
package com.phicdy.mycuration.rss;

import com.phicdy.mycuration.db.DatabaseAdapter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class UnreadCountManager {
    private int total = 0;
    private final Map<Integer, Integer> unreadCountMap = new HashMap<>();
    private ArrayList<Feed> allFeeds;
    private final DatabaseAdapter adapter;
    private static UnreadCountManager mgr;

    private UnreadCountManager() {
        this.adapter = DatabaseAdapter.getInstance();
        init();
    }

    public static UnreadCountManager getInstance() {
        if (mgr == null) {
            synchronized (UnreadCountManager.class) {
                if (mgr == null) {
                    mgr = new UnreadCountManager();
                }
            }
        }
        return mgr;
    }

    private void init() {
        total = 0;
        allFeeds = adapter.getAllFeedsWithNumOfUnreadArticles();
        synchronized (unreadCountMap) {
            unreadCountMap.clear();
            for (Feed feed : allFeeds) {
                unreadCountMap.put(feed.getId(), feed.getUnreadAriticlesCount());
                total += feed.getUnreadAriticlesCount();
            }
        }
    }

    public void addFeed(Feed feed) {
        if (feed == null) {
            return;
        }
        unreadCountMap.put(feed.getId(), feed.getUnreadAriticlesCount());
        total += feed.getUnreadAriticlesCount();
    }

    public void deleteFeed(int feedId) {
        if (!unreadCountMap.containsKey(feedId)) {
            return;
        }
        total -= unreadCountMap.get(feedId);
        unreadCountMap.remove(feedId);
    }

    public void conutUpUnreadCount(int feedId) {
        synchronized (unreadCountMap) {
            if (!unreadCountMap.containsKey(feedId)) {
                return;
            }
            int count = unreadCountMap.get(feedId) + 1;
            unreadCountMap.put(feedId, count);
        }
        total++;
        updateDatbase(feedId);
    }

    public void countDownUnreadCount(int feedId) {
        synchronized (unreadCountMap) {
            if (!unreadCountMap.containsKey(feedId)) {
                return;
            }
            int count = unreadCountMap.get(feedId);
            if (count == 0) {
                return;
            }
            unreadCountMap.put(feedId, --count);
        }
        total--;
        updateDatbase(feedId);
    }

    public int getUnreadCount(int feedId) {
        return unreadCountMap.containsKey(feedId) ? unreadCountMap.get(feedId) : -1;
    }

    public int getTotal() {
        return total;
    }

    private void updateDatbase(final int feedId) {
        if (!unreadCountMap.containsKey(feedId)) {
            return;
        }

        synchronized (unreadCountMap) {
            final int count = unreadCountMap.get(feedId);
            new Thread() {
                @Override
                public void run() {
                    adapter.updateUnreadArticleCount(feedId, count);
                }
            }.start();
        }
    }

    public void readAll(int feedId) {
        synchronized (unreadCountMap) {
            if (!unreadCountMap.containsKey(feedId)) {
                return;
            }
            int count = unreadCountMap.get(feedId);
            total -= count;
            unreadCountMap.put(feedId, 0);
        }
        updateDatbase(feedId);
    }

    public void readAll() {
        synchronized (unreadCountMap) {
            total = 0;
            for (Feed feed : allFeeds) {
                unreadCountMap.put(feed.getId(), 0);
                updateDatbase(feed.getId());
            }
        }
    }

    public void refreshConut(int feedId) {
        synchronized (unreadCountMap) {
            // Decrease original unread count
            if (!unreadCountMap.containsKey(feedId)) {
                return;
            }
            int oldCount = unreadCountMap.get(feedId);
            total -= oldCount;

            // Calc unread count from database
            int count = adapter.getNumOfUnreadArtilces(feedId);
            adapter.updateUnreadArticleCount(feedId, count);
            unreadCountMap.put(feedId, count);
            total += count;
        }
    }

    public int getCurationCount(int curationId) {
        return adapter.calcNumOfAllUnreadArticlesOfCuration(curationId);
    }
}

File: mycuration/src/main/java/com/phicdy/mycuration/domain/rss/UnreadCountManager.kt
package com.phicdy.mycuration.domain.rss

import android.support.annotation.VisibleForTesting
import com.phicdy.mycuration.data.db.DatabaseAdapter
import com.phicdy.mycuration.data.rss.Feed

object UnreadCountManager {
    var total = 0
        private set
    private val unreadCountMap = hashMapOf<Int, Int>()
    private var adapter: DatabaseAdapter

    @VisibleForTesting
    fun inject(adapter: DatabaseAdapter) {
        this.adapter = adapter
    }

    init {
        total = 0
        adapter = DatabaseAdapter.getInstance()
        val allFeeds = adapter.allFeedsWithNumOfUnreadArticles
        synchronized(unreadCountMap) {
            unreadCountMap.clear()
            allFeeds.forEach {
                unreadCountMap[it.id] = it.unreadAriticlesCount
                total += it.unreadAriticlesCount
            }
        }
    }

    fun clear() {
        total = 0
        unreadCountMap.clear()
    }

    fun addFeed(feed: Feed?) {
        if (feed == null) {
            return
        }
        unreadCountMap[feed.id] = feed.unreadAriticlesCount
        total += feed.unreadAriticlesCount
    }

    fun deleteFeed(feedId: Int) {
        if (!unreadCountMap.containsKey(feedId)) return
        total -= unreadCountMap[feedId]!!
        unreadCountMap.remove(feedId)
    }

    fun conutUpUnreadCount(feedId: Int) {
        synchronized(unreadCountMap) {
            if (!unreadCountMap.containsKey(feedId)) return
            val count = unreadCountMap[feedId]!! + 1
            unreadCountMap.put(feedId, count)
        }
        total++
        updateDatbase(feedId)
    }

    fun countDownUnreadCount(feedId: Int) {
        synchronized(unreadCountMap) {
            if (!unreadCountMap.containsKey(feedId)) return
            unreadCountMap[feedId]!!.also { count ->
                if (count == 0) return
                unreadCountMap[feedId] = count - 1
            }
        }
        total--
        updateDatbase(feedId)
    }

    fun getUnreadCount(feedId: Int): Int {
        return if (unreadCountMap.containsKey(feedId)) unreadCountMap[feedId]!! else -1
    }

    private fun updateDatbase(feedId: Int) {
        if (!unreadCountMap.containsKey(feedId)) return

        synchronized(unreadCountMap) {
            val count = unreadCountMap[feedId]!!
            object : Thread() {
                override fun run() {
                    adapter.updateUnreadArticleCount(feedId, count)
                }
            }.start()
        }
    }

    fun readAll(feedId: Int) {
        synchronized(unreadCountMap) {
            if (!unreadCountMap.containsKey(feedId)) return
            val count = unreadCountMap[feedId]!!
            total -= count
            unreadCountMap.put(feedId, 0)
        }
        updateDatbase(feedId)
    }

    fun readAll() {
        synchronized(unreadCountMap) {
            total = 0
            for (id in unreadCountMap.keys) {
                unreadCountMap[id] = 0
                updateDatbase(id)
            }
        }
    }

    fun refreshConut(feedId: Int) {
        synchronized(unreadCountMap) {
            // Decrease original unread count
            if (!unreadCountMap.containsKey(feedId)) return
            val oldCount = unreadCountMap[feedId]!!
            total -= oldCount

            // Calc unread count from database
            val count = adapter.getNumOfUnreadArtilces(feedId)
            adapter.updateUnreadArticleCount(feedId, count)
            unreadCountMap[feedId] = count
            total += count
        }
    }

    fun getCurationCount(curationId: Int): Int {
        return adapter.calcNumOfAllUnreadArticlesOfCuration(curationId)
    }
}
